qiskit/algorithms/eigensolvers/vqd.py

```diff
@@ -0,0 +1,477 @@
+
+        Args:
+            step: level of energy being calculated. 0 for ground, 1 for first excited state...
+            operator: The operator whose energy to evaluate.
+            prev_states: List of parameters from previous rounds of optimization.
+
+        Returns:
+            A callable that computes and returns the energy of the hamiltonian
+            of each parameter.
+
+        Raises:
+            AlgorithmError: If the circuit is not parameterized (i.e. has 0 free parameters).
+            AlgorithmError: If operator was not provided.
+            RuntimeError: If the previous states array is of the wrong size.
+        """
+
+        num_parameters = self.ansatz.num_parameters
+        if num_parameters == 0:
+            raise AlgorithmError("The ansatz must be parameterized, but has no free parameters.")
+
+        if step > 1 and (len(prev_states) + 1) != step:
+            raise RuntimeError(
+                f"Passed previous states of the wrong size."
+                f"Passed array has length {str(len(prev_states))}"
+            )
+
+        self._check_operator_ansatz(operator)
+
+        prev_circs = []
+        for state in range(step - 1):
+            prev_circs.append(self.ansatz.bind_parameters(prev_states[state]))
+
+        def evaluate_energy(parameters: np.ndarray) -> np.ndarray | float:
+
+            try:
+                estimator_job = self.estimator.run(
+                    circuits=[self.ansatz], observables=[operator], parameter_values=[parameters]
+                )
+                estimator_result = estimator_job.result()
+                values = estimator_result.values
+
+            except Exception as exc:
+                raise AlgorithmError("The primitive job to evaluate the energy failed!") from exc
+
+            if step > 1:
+                # Compute overlap cost
+                fidelity_job = self.fidelity.run(
+                    [self.ansatz] * len(prev_circs),
+                    prev_circs,
+                    [parameters] * len(prev_circs),
+                    [prev_states[:-1]],
+                )
+                costs = fidelity_job.result().fidelities
+
+                for (state, cost) in zip(range(step - 1), costs):
+                    values += np.real(betas[state] * cost)
+
+            if self.callback is not None:
+                metadata = estimator_result.metadata
+                for params, value, meta in zip([parameters], values, metadata):
+                    self._eval_count += 1
+                    self.callback(self._eval_count, params, value, meta, step)
+            else:
+                self._eval_count += len(values)
+
+            return values if len(values) > 1 else values[0]
+
+        return evaluate_energy
+
+    @staticmethod
+    def _build_vqd_result() -> VQDResult:
+
+        result = VQDResult()
+        result.optimal_point = []
+        result.optimal_parameters = []
+        result.optimal_value = []
```

Comment:
Though eigenvalues is plural, should we have others this way too - optimizer_times, optimal_values etc. They were singular in VQE, which has just a single value, but here arguably I think they should all be plural (ie. with s on the end)